### PR TITLE
chore: created and tested agent access entity

### LINKED ISSUE #25: Define Entity AgentAccess
#### Overview:
`AgentAccess` is a specialized access entity that allows a designated agent to act on behalf of the account holder in agency-related processes. This role enables the agent to perform specific actions that the account holder would typically handle, but it does not provide the authority to manage or delegate access to other users. `AgentAccess` is primarily designed to facilitate tasks where the account holder cannot be directly involved but needs representation, such as in business transactions, legal processes, or financial operations.

#### Key Characteristics:

1. **Impersonation of the Account Holder:**
   - `AgentAccess` enables the agent to act as a proxy for the account holder, allowing them to perform actions in agency processes where the account holder is not available or prefers to delegate certain tasks.
   - This impersonation is limited to predefined activities or processes that have been authorized by the account holder.
   - The agent’s actions are carried out "in the name of" the account holder, meaning that the account holder remains responsible for any decisions or transactions made by the agent within the scope of their permissions.

2. **Limited Scope of Permissions:**
   - `AgentAccess` grants the agent the ability to execute specific actions on behalf of the account holder but restricts them from making changes to the account's access permissions.
   - Typical permissions for an agent might include:
     - **Executing Payments or Transfers:** The agent can initiate payments, transfers, or transactions on behalf of the account holder, but they cannot delegate this authority to others.
     - **Managing Specific Operations:** The agent might be allowed to handle certain types of administrative tasks (e.g., filing taxes, conducting transactions, fulfilling business obligations) under the authority of the account holder.
   - The exact permissions are defined by the `scope` field of the `AgentAccess` entity, ensuring that the agent can only perform tasks authorized by the account holder.

3. **No Management of Access:**
   - Unlike `ManagerAccess`, the agent is not permitted to manage or modify access to the account for other users.
   - They cannot grant, modify, or revoke access for other roles (e.g., AuditorAccess, PoAAccess), ensuring that the agent’s authority is restricted to actions on behalf of the account holder.
   - This limitation helps ensure that the agent's role remains focused on execution rather than governance of the account.

4. **Delegated Authority:**
   - `AgentAccess` is granted explicitly by the account holder (`HolderAccess`) or by a manager with sufficient authority to delegate such a role.
   - The delegation of `AgentAccess` may be temporary (e.g., for a specific transaction or period) or ongoing, depending on the needs of the account holder.
   - The account holder retains full control and responsibility over the actions taken by the agent and can revoke this access at any time.

5. **Weight:**
   - The `weight` of `AgentAccess` typically reflects partial authority, often lower than that of a full account holder. For example:
     - A `weight` of **0.5** might signify that the agent has significant authority to act on behalf of the account holder in certain processes but is not fully equivalent to the account holder.
     - The agent’s `weight` indicates their relative authority compared to other access roles, especially if multiple agents are involved or if their powers are shared.

6. **Status:**
   - The status of `AgentAccess` can vary depending on the current relationship between the account holder and the agent:
     - **Active:** The agent is fully authorized to act on behalf of the account holder.
     - **Restricted:** The agent’s permissions may be temporarily limited (e.g., due to changes in legal conditions or the account holder’s preferences).
     - **Suspended:** The agent can no longer act on behalf of the account holder, typically following the revocation of their authority.
   - These status changes can occur automatically (e.g., when the agent’s role expires) or manually at the discretion of the account holder or an authorized manager.

7. **Accountability and Logging:**
   - Every action taken by an agent is logged and traceable to ensure transparency and accountability.
   - All transactions or processes initiated by the agent are recorded as being executed under the authority of the account holder.
   - Audit logs for `AgentAccess` include details of the operations performed by the agent, ensuring that any potential misuse or errors can be traced and attributed accurately.

8. **Conditions on Access:**
   - The account holder or manager may impose specific conditions on the agent’s access, such as:
     - **Monetary Limits:** The agent may only execute transactions up to a certain limit.
     - **Time Constraints:** The agent’s access may be limited to a specific period or project (e.g., acting as a proxy during a business trip).
     - **Geographical or Process Constraints:** The agent’s authority might be restricted to certain types of transactions or geographical regions, ensuring that their power is limited to predefined activities.

#### Example Workflow for AgentAccess:

1. **Executing a Payment on Behalf of the Holder:**
   - The account holder authorizes an agent to pay invoices or transfer funds to a business partner while they are away on vacation.
   - The agent uses their `AgentAccess` to execute the payments as specified by the account holder. They cannot, however, modify any other user’s access or make changes to the account settings.

2. **Handling Tax Filing:**
   - A business account holder grants an accountant `AgentAccess` to handle tax filings on their behalf.
   - The accountant can gather the necessary financial information and submit the taxes, but they are unable to view sensitive account details not directly related to the filing or manage access to the account.

3. **Legal Representation in Financial Matters:**
   - A legal representative is granted `AgentAccess` to execute financial transactions as part of a legal settlement. They can withdraw funds, make payments, or fulfill other legal obligations but cannot assign access to anyone else or alter the account structure.

#### Key Considerations for AgentAccess:

- **Clear Scope Definition:** The account holder should carefully define the scope of the agent’s permissions to ensure that they only perform tasks that are strictly necessary. Overly broad permissions could result in unintended actions or misuse.
- **Accountability:** Since the agent acts on behalf of the account holder, there must be clear and detailed logging of all actions taken under `AgentAccess` to maintain transparency and protect both the agent and the account holder.
- **Revocability:** The account holder must have the ability to quickly and easily revoke or suspend the agent’s access if their role is no longer required or if any misuse is suspected.

#### Scenarios for Suspension or Restriction:

- **Completion of a Task:** Once an agent has completed their assigned task (e.g., executing a business deal or paying off a loan), the account holder may choose to suspend or revoke their `AgentAccess` to prevent any further actions.
- **Legal or Regulatory Concerns:** If there are legal reasons to suspend the agent’s role (e.g., an investigation or court order), their access can be restricted or revoked immediately.
- **Account Holder Decision:** The account holder can revoke `AgentAccess` at any time, such as if they no longer trust the agent or if the agent’s services are no longer needed.

#### Example Use Cases:

- **Corporate Representation:** A company might grant `AgentAccess` to a legal representative or employee to handle specific financial operations, such as paying suppliers or fulfilling contractual obligations, without allowing them full access to the company's bank account.
- **Business Travel:** An account holder traveling abroad may grant `AgentAccess` to a trusted colleague or assistant to manage financial matters in their absence, ensuring that important transactions can still occur without direct involvement.
- **Real Estate Transactions:** A property owner might grant `AgentAccess` to a real estate agent to handle payments or financial aspects of a property sale, acting on the owner’s behalf during negotiations or settlements.

### PR DESCRIPTION
**Pull request by Koufan-De-King**
Created

- agentaccess class,
- a repository for it's eventual connection to the database
and wrote a basic test using JUnit.

### CODE REVIEW

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AccessScope.java [hunk 1]
```diff
@@ -11,4 +11,16 @@ public enum AccessScope {
...
+            try {
+                AccessScope requestedAction = AccessScope.valueOf(action);
+                return this == requestedAction;
+            } catch (IllegalArgumentException e) {
+                // Action not allowed
+                return false;
+            }
+        }
+    
+
```
Comment (NkwaTambe) [inline ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AccessScope.java]: This is an `Enum class` 

These implementation should not be here..

**An enum is a special "class" that represents a group of constants (unchangeable variables, like final variables)**

as a **good practice** , no implementation should be done in the `Enum class`

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AgentAccess.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    // Example usage scenario: Execute a payment on behalf of the account holder
+    public void executePayment(double amount) {
+        if (this.getStatus() == AccessStatus.ACTIVE && this.getScope().allowsAction("EXECUTE_PAYMENT")) {
+           return;
+            // Perform the payment logic
+        } else {
+            throw new IllegalStateException("Payment execution not allowed under the current access status or scope.");
+        }
+    }
+}
```
Comment (NkwaTambe) [inline ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AgentAccess.java]: Perfect @Koufan-De-King , i love the inheritance implementation here..

please meet @nancymuyeh  and explain to her your idea in implementing these class..

she will need some understanding for her class implementation...

i am focus on the `inheritance` , the **BankAccountAcces Entity** serves as an entery for all `fields` definitions

File: ledgers-user-management/ledgers-user-management-service-api/src/main/java/de/adorsys/ledgers/um/api/service/UserService.java [hunk 3]
```diff
@@ -7,11 +7,18 @@
...
+import de.adorsys.ledgers.um.api.domain.UserExtendedBO;
+import de.adorsys.ledgers.um.api.domain.UserRoleBO;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Map;
 
 public interface UserService {
```
Comment (NkwaTambe) [inline ledgers-user-management/ledgers-user-management-service-api/src/main/java/de/adorsys/ledgers/um/api/service/UserService.java:23]: These is a different module, it should not be here.

looks like you deleted things..

Please fix these.

**Review by NkwaTambe — CHANGES_REQUESTED** [review]

**Comment by NkwaTambe** [comment]
Hello @Koufan-De-King 

please for your implementation, for it to be complete as per the ticket description :

`**Accountability and Logging:**
Every action taken by an agent is logged and traceable to ensure transparency and accountability.
All transactions or processes initiated by the agent are recorded as being executed under the authority of the account holder.
Audit logs for AgentAccess include details of the operations performed by the agent, ensuring that any potential misuse or errors can be traced and attributed accurately.`

you have to implement the above.

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AgentAccess.java (hunk 2 above)
Comment (NkwaTambe) [inline ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AgentAccess.java]: Good job @Koufan-De-King 

I think you are missing some things  in here..

please read the ticket description again thanks.

File: ledgers-user-management/ledgers-user-management-service-api/src/main/java/de/adorsys/ledgers/um/api/service/UserService.java (hunk 3 above)
Comment (Koufan-De-King) [reply ledgers-user-management/ledgers-user-management-service-api/src/main/java/de/adorsys/ledgers/um/api/service/UserService.java:23]: This was necessary to correct a failing pmd check.

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AgentAccess.java (hunk 2 above)
Comment (Koufan-De-King) [reply ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AgentAccess.java]: Will be taken care of @NkwaTambe !

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AccessScope.java (hunk 1 above)
Comment (Koufan-De-King) [reply ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/AccessScope.java]: Understood.

**Review by nancymuyeh — APPROVED** [review]

**Review by NkwaTambe — APPROVED** [review]

**Pull request merged**